samza-azure/src/main/java/org/apache/samza/coordinator/AzureJobCoordinator.java

```diff
@@ -443,18 +440,24 @@ private void onNewJobModelAvailable(final String nextJMVersion) {
   private void onNewJobModelConfirmed(final String nextJMVersion) {
     LOG.info("pid=" + processorId + "new version " + nextJMVersion + " of the job model got confirmed");
 
-    // Delete previous value
-    if (table.getEntity(currentJMVersion.get(), processorId) != null) {
-      table.deleteProcessorEntity(currentJMVersion.get(), processorId);
-    }
-    if (table.getEntity(INITIAL_STATE, processorId) != null) {
-      table.deleteProcessorEntity(INITIAL_STATE, processorId);
-    }
+    String prevVersion = currentJMVersion.get();
 
-    //Start heartbeating to new entry only when barrier reached.
+    //Start heart-beating to new entry only when barrier reached.
     //Changing the current job model version enables that since we are heartbeating to a row identified by the current job model version.
     currentJMVersion.getAndSet(nextJMVersion);
 
+    // Delete previous value
+    ProcessorEntity entity = table.getEntity(prevVersion, processorId);
+    if (entity != null) {
+      entity.setEtag("*");
+      table.deleteProcessorEntity(entity);
```

Comment:
Please correct me if i'm wrong here.

The heart beat thread block could still be executed before this block gets triggered trying to delete the azure table row resulting in the same exception mentioned in the JIRA. Since there're no synchronization/communication between these two regions(thread), I think there're no ordering guarantees.

I think it's much simpler to remove deletes from all the followers and do the GC only on the leader itself based upon a processor liveness timeout configuration(similar to zk session timeout). The followers will simply join with creating an entry and will not delete on exit(leader will do the GC(delete all the processors from previous generations) on each job model generation or once when a leader gets elected).

If it's essential it could be changed as a part of this JIRA or add a todo with JIRA for later(since it happens during stream-processor shutdown).

Reply:
@shanthoosh You are right in that there are no ordering guarantees here. Setting the etag to "*" only forces deletion irrespective of another thread accessing the table. So, it doesn't solve the synchronization problem. For now, it simply prevents the processor from throwing up exceptions due to locking when it is shutting down. 

> I think it's much simpler to remove deletes from all the followers and do the GC only on the leader itself based upon a processor liveness timeout configuration(similar to zk session timeout).
Yes. That is the solution I would like to see as well. However, given how we have partitioned the tables today, it seems like a very expensive operation. I will create a JIRA as a follow-up for this to fix the synchronization issues and clean-up of the tables.